### PR TITLE
widgets: add Image

### LINKED ISSUE #10: [Idea] Adding a user picture
Read the image stored at ~/.face or /var/lib/Accountservice/users/[username] and render that image on top of  the lable on a rounded frame or define the picture in the config file either.

### PR DESCRIPTION
**Pull request by bvr-yr**
that's a simple implementation, just to draw a circle with picture
mostly copy-pasted from background widget and input-field :p

close #10

055708-output- eDP-1

060544-output- eDP-1

```ini
avatar {
    border_color = 0xffdddddd
    border_size = 4
    size = 150
    position = 0, 200
    halign = center
    valign = center
}
```

&check; Shadowable

### CODE REVIEW

**Review by vaxerski — COMMENTED** [review]
generally nice work, but I'd add a `rounding` option for customizing the rounding.

Can be default `-1` for "full circle"

**Comment by bvr-yr** [comment]
hah, yeah initially i was doing exactly that and almost committed, but then tried image with non 1:1 ratio
without touching `renderTexture` i could get
- either 1:1 image but stretched texture
- ore ok texture but not 1:1 avatar

with function like that, when i try to allow any number except `size / 2` i get smth like

image

if ratio is not 1:1
maybe i am missing smth?

**Comment by bvr-yr** [comment]
basically, i can check for that and use rounding only for 1:1, and mention it in the Wiki

**Comment by bvr-yr** [comment]
i found it :)

**Comment by bvr-yr** [comment]
hm, it literally can be used just like `Picture` widget, maybe make it so and add `avatar` option to crop middle since it works now?

**Comment by vaxerski** [comment]
yea thats ugly, dont change renderTexture, not needed.

**Comment by bvr-yr** [comment]
but now it works `true` is passed, or it is not ok?

**Comment by vaxerski** [comment]
dont edit the texture renderer

**Comment by bvr-yr** [comment]
even just like so?
```diff
-void CRenderer::renderTexture(const CBox& box, const CTexture& tex, float a, int rounding, std::optional<wl_output_transform> tr) {
+void CRenderer::renderTexture(const CBox& box, const CTexture& tex, float a, int rounding, std::optional<wl_output_transform> tr, bool avatar, const CBox& avabox) {
```
```diff
-    const auto TOPLEFT  = Vector2D(box.x, box.y);
-    const auto FULLSIZE = Vector2D(box.width, box.height);
+    const auto TOPLEFT  = avatar ? Vector2D(avabox.x, avabox.y) : Vector2D(box.x, box.y);
+    const auto FULLSIZE = avatar ? Vector2D(avabox.width, avabox.height) : Vector2D(box.width, box.height);
```
just to clarify, i'm totally ok with that

**Comment by bvr-yr** [comment]
nvm i agree with you, its not that hard to prepare an image if you want an avatar

@fufexan check please if i done nix right

File: nix/hm-module.nix [hunk 1]
```diff
@@ -162,6 +162,77 @@ in {
...
+          monitor = mkOption {
+            description = "The monitor to draw an image";
+            type = str;
+            default = "";
+          };
+
+          path = mkOption {
+            description = "The path to source image";
+            type = str;
+            default = "echo '/home/me/cutie.png'"; # only png supported for now
```
Comment (fufexan) [inline nix/hm-module.nix:178]: Why not just 
```suggestion
            default = "/home/me/cutie.png"; # only png supported for now
```

**Review by fufexan — COMMENTED** [review]
I have a few things to point out:
- the default value for rounding can be 0 instead of -1
- the widget should be renamed `image`, as not every image is a picture
- rename `frame` to `border`, to be more familiar to people who know CSS

File: nix/hm-module.nix (hunk 1 above)
Comment (bvr-yr) [reply nix/hm-module.nix:178]: i also was surprised, just copy-pasted from background, thought this is intended

**Review by vaxerski — APPROVED** [review]
lgtm for code part

File: nix/hm-module.nix [hunk 2]
```diff
@@ -162,6 +162,77 @@ in {
...
+            default = 150;
+          };
+
+          rounding = mkOption {
+            description = "The rounding of the image";
+            type = int;
+            default = -1;
+          };
+
+          frame_size = mkOption {
```
Comment (fufexan) [inline nix/hm-module.nix:193]: This is now `border`, I think you forgot to change it.

Comment (bvr-yr) [reply nix/hm-module.nix]: oops true. done

**Review by fufexan — APPROVED** [review]
yup, looks good

**Comment by bvr-yr** [comment]
yeah truly, @fufexan probably forgot about pending Wiki [MR](https://github.com/hyprwm/hyprland-wiki/pull/529) when merged this

**Comment by fufexan** [comment]
Yeah, sorry.

**Comment by alba4k** [comment]
Unrelated, but @bvr-yr people are asking for your wallpaper, maybe check that out :shrug: 

https://discord.com/channels/961691461554950145/1215490910364762143/1219325620497158174

**Comment by bvr-yr** [comment]
discord refuses to send me sms, probably because Belarus is banned? (lul)
@vaxerski what do?

**Comment by fufexan** [comment]
@bvr-yr I don't know why it wouldn't let you send me a DM, I have them open.

Anyway, found you in the members list and gave you permission.

**Pull request merged**